lib/puppet/transaction/report.rb

```diff
@@ -240,6 +244,46 @@ def name
     host
   end
 
+  # Provide a summary of the changes
+  def displaychanges
+      ret = ""
+
+      # display the different changes for each type
+      changes.keys.sort.each { |type|
+          ret += "%s:\n" % type
+          changes[type].keys.sort.each { |event|
+              # Count the number of occurences for each value in the changes[type][event]
+              count = changes[type][event].inject(Hash.new(0)) {|h,i| h[i] += 1; h }
+              ret += count.to_a.collect { |status, nb|
+                  if nb > 1
+                      "   %15s %s (%s)" % [status + ":", event, nb]
+                  else
+                      "   %15s %s" % [status + ":", event]
+                  end
+              }.join("\n")
+              ret += "\n"
+          }
+      }
+      return ret
+  end
+
+   # Copy events in a structure that can be return with the report
+   def sumup()
+       changes= {}
+       resource_statuses.each do |name, status|
+           status.events.each do |e|
+               e.resource =~ /([^\[]*)\[([^\]]*)\]/
+               type = $1
+               title = $2
```

Comment:
I'm fairly certain all events for a status, are associated with the same resource, and you can get the type and title of that resource from the `status` object itself, avoiding this regex.